### PR TITLE
Avoid accessing private loc methods

### PR DESCRIPTION
**Pull request by jbrockmendel**
Works toward fixing #27344 

Adds types where unambiguous

### CODE REVIEW

**Comment by jbrockmendel** [comment]
@WillAyd can you help me make heads or tails out of the mypy complaint?

**Comment by WillAyd** [comment]
Yea I imagine it's because of part of Lin e 1745:

```python
[key] + [slice(None)]
```

So the left operand is inferred to `List[str]` while the right hand side is a `slice` and Mypy is complaining that you are appending a slice to a list of what maybe should be just str.

Could either:

1. Create a new variable for this typed to `List[Union[str, slice]]`
2. # type: ignore this line

I think option 1 is better since this code is shadowing `key` variable anyway with a different type

**Comment by jbrockmendel** [comment]
> Yea I imagine it's because of part of Line 1745:

The part that I don't get is why is it only complaining now?  The types being added should be a strict improvement typing-wise, right?

**Comment by WillAyd** [comment]
Haven’t reviewed in detail but if first annotations in module Mypy will actually start checking. Before may have just skipped

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -222,7 +223,7 @@ def _validate_key(self, key, axis: int):
         """
         raise AbstractMethodError(self)
 
-    def _has_valid_tuple(self, key):
+    def _has_valid_tuple(self, key: tuple):
```
Comment (WillAyd) [inline pandas/core/indexing.py]: Can you use the generic `Tuple` from the typing module? Providing the type(s) contained within would be even better (maybe `Tuple[Hashable]`?)

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -1739,22 +1736,26 @@ def _getitem_scalar(self, key):
...
+    def _get_partial_string_timestamp_match_key(self, key, labels: Index):
         """Translate any partial string timestamp matches in key, returning the
         new key (GH 10331)"""
         if isinstance(labels, MultiIndex):
             if isinstance(key, str) and labels.levels[0].is_all_dates:
                 # Convert key '2016-01-01' to
                 # ('2016-01-01'[, slice(None, None, None)]+)
-                key = tuple([key] + [slice(None)] * (len(labels.levels) - 1))
+                key = tuple(
+                    [key] + [slice(None)] * (len(labels.levels) - 1)
```
Comment (WillAyd) [inline pandas/core/indexing.py]: The more I think about this a better idiom would be something like:

```python
new_key = None # type: List[str, slice]
if isinstance(labels, MultiIndex):
   ...
   new_key = tuple([key] + [slice(None)] * len(labels.levels) - 1)
    ...
```

Should appease mypy and prevent the shadowing of the `key` variable

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py [hunk 3]
```diff
@@ -3484,7 +3484,9 @@ def _setitem_array(self, key, value):
                 for k1, k2 in zip(key, value.columns):
                     self[k1] = value[k2]
             else:
-                indexer = self.loc._convert_to_indexer(key, axis=1)
+                indexer = self.loc._get_listlike_indexer(
```
Comment (jreback) [inline pandas/core/frame.py]: ```self.columns.get_indexer(key)```?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/frame.py]: Looks like this returns something different, will have to keep digging

**Comment by jbrockmendel** [comment]
gentle ping; there are (requested) follow-ups

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**